src/components/mdx/Code.tsx

```diff
@@ -0,0 +1,175 @@
+import React, {FunctionComponent, useEffect, useState} from 'react'
+import styled from '@emotion/styled'
+import Highlight, {defaultProps} from 'prism-react-renderer'
+import theme from 'prism-react-renderer/themes/nightOwl'
+
+const RE = /{([\d,-]+)}/
+
+function calculateLinesToHighlight(meta: string) {
+  if (RE.test(meta)) {
+    const lineNumbers = RE.exec(meta)![1]
+      .split(',')
+      .map((v) => v.split('-').map((y) => parseInt(y, 10)))
+    return (index: number) => {
+      const lineNumber = index + 1
+      const inRange = lineNumbers.some(([start, end]) =>
+        end ? lineNumber >= start && lineNumber <= end : lineNumber === start,
+      )
+      return inRange
+    }
+  }
+  return () => false
+}
+
+type CodeProps = {
+  className: string
+  metastring: string
+  children: string
+}
+
+const Code: FunctionComponent<CodeProps> = ({
+  children,
+  className,
+  metastring,
+}) => {
+  const language = className ? className.replace(/language-/, '') : ''
+  const shouldHighlightLine = calculateLinesToHighlight(metastring)
+
+  const defaultCopyText = 'Copy'
+  const [copyText, setCopyText] = useState(defaultCopyText)
+
+  useEffect(() => {
+    let current = true
+    if (copyText !== defaultCopyText) {
+      setTimeout(() => {
+        if (current) {
+          setCopyText(defaultCopyText)
+        }
+      }, 2000)
+    }
+    return () => {
+      current = false
+    }
+  }, [copyText])
+
+  function copy(event: {preventDefault: () => void}) {
+    event.preventDefault()
+    navigator.clipboard.writeText(children).then(
```

Comment:
This copies the whole code block. It uses the Clipboard interface's `writeText()` property to write the text string to the system clipboard.

Reply:
https://user-images.githubusercontent.com/57044804/103181441-e36e2a80-4855-11eb-9908-9bb80724bc54.mov

Reply:
thought: there's also `useCopyToClipboard` hook available from `react-use` lib which might have better browser/os support. [example from lesson page](https://github.com/eggheadio/egghead-next/blob/main/src/components/pages/lessons/LessonInfo.tsx#L6)